test/cpp_api_parity/module_impl_check.py

```diff
@@ -0,0 +1,308 @@
+  // to set the RNG seed manually.
+  torch::manual_seed(0);
+
+  // Forward pass
+  auto cpp_output = module(${cpp_forward_args_symbols});
+
+  // Save the output into a file to be compared in Python later
+  write_ivalue_to_file(torch::IValue(cpp_output), forward_output_file_path);
+
+  // Backward pass
+  cpp_output.sum().backward();
+
+  // Put all gradients into a c10::Dict, save it into a file to be compared in Python later
+  c10::Dict<std::string, torch::Tensor> grad_dict;
+  for (const auto& param : module->named_parameters()) {
+    torch::Tensor grad = param.value().grad();
+    if (grad.is_sparse()) {
+      grad_dict.insert(param.key() + "_grad_indices", grad.coalesce().indices());
+      grad_dict.insert(param.key() + "_grad_values", grad.coalesce().values());
+    } else {
+      grad_dict.insert(param.key() + "_grad", grad);
+    }
+  }
+
+  write_ivalue_to_file(torch::IValue(grad_dict), backward_grad_dict_file_path);
+}
+""")
+
+def run_python_forward_backward(unit_test_class, test_params):
+    device = test_params.device
+    module = test_params.test_instance.constructor(*test_params.test_instance.constructor_args).to(device)
+
+    inputs = set_python_tensors_requires_grad(move_python_tensors_to_device(
+        [arg_value for _, arg_value in test_params.arg_dict['input']], device))
+    inputs = inputs + move_python_tensors_to_device(
+        [arg_value for _, arg_value in test_params.arg_dict['target']], device)
+    inputs = inputs + move_python_tensors_to_device(
+        [arg_value for _, arg_value in test_params.arg_dict['extra_args']], device)
+
+    # Some modules (such as `RReLU`) create random tensors in their forward pass.
+    # To make sure the random tensors created are the same in Python/C++, we need
+    # to set the RNG seed manually.
+    torch.manual_seed(0)
+
+    # Forward pass
+    python_output = module(*inputs)
+
+    # NOTE: This is a workaround to allow any module to be traced.
+    # We can do this because we are only interested in transferring
+    # the Python module's parameters and buffers to the C++ module.
+    module.forward = types.MethodType(lambda self, input: input, module)
+    script_module = torch.jit.trace(module, torch.tensor(0))
+
+    # Backward pass
+    python_output.sum().backward()
+
+    # Put all gradients into a dict, to be compared later
+    python_grad_dict = {}
+    for name, param in module.named_parameters():
+        grad = param.grad
+        if grad.is_sparse:
+            python_grad_dict[name + "_grad_indices"] = grad.coalesce().indices()
+            python_grad_dict[name + "_grad_values"] = grad.coalesce().values()
+        else:
+            python_grad_dict[name + "_grad"] = grad
+
+    return script_module, python_output, python_grad_dict
+
+def test_forward_backward(unit_test_class, test_params):
+    module_variant_name = test_params.module_variant_name
+    cpp_tmp_folder = test_params.cpp_tmp_folder
```

Comment:
Nit: it's nice to remove the cpp_tmp_folder so that we can be more sure that a previous run of the test didn't affect the current run